programs/mango-v4/src/state/token_conditional_swap.rs

```diff
@@ -0,0 +1,189 @@
+use anchor_lang::prelude::*;
+
+use derivative::Derivative;
+use fixed::types::I80F48;
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use static_assertions::const_assert_eq;
+use std::mem::size_of;
+
+use crate::state::*;
+
+#[derive(
+    Eq,
+    PartialEq,
+    Copy,
+    Clone,
+    TryFromPrimitive,
+    IntoPrimitive,
+    Debug,
+    AnchorSerialize,
+    AnchorDeserialize,
+)]
+#[repr(u8)]
+pub enum TokenConditionalSwapPriceThresholdType {
+    PriceOverThreshold,
+    PriceUnderThreshold,
+}
+
+#[zero_copy]
+#[derive(AnchorDeserialize, AnchorSerialize, Derivative, bytemuck::Pod)]
+#[derivative(Debug)]
+pub struct TokenConditionalSwap {
+    pub id: u64,
+
+    /// maximum amount of native tokens to buy or sell
+    pub max_buy: u64,
+    pub max_sell: u64,
+
+    /// how many native tokens were already bought/sold
+    pub bought: u64,
+    pub sold: u64,
+
+    /// timestamp until which the conditional swap is valid
+    pub expiry_timestamp: u64,
```

Comment:
maybe is_active method everywhere should also use this field to mark as active, not active or free tcs slot

Reply:
maybe I'll rename `is_active()` to `has_data` or something